lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -25,54 +26,86 @@ def __init__(self, context):
     def execute(self) -> int:
         errors = 0
 
-        if self.verify_tool_group(self.context.group) != 0:
-            return 1
+        groups = self.context.group.split(",")
+        # groups is never empty and if the user doesn't specify a --group,
+        # then it has the value "default"
+        for group in groups:
+            if self.verify_tool_group(group) != 0:
+                self.logger.warn(f'No such group "{group}".')
+                errors = 1
+                continue
 
-        try:
+            errors, tools_not_found = self._clear_remotes(group)
+
+            if self.context.name and tools_not_found:
+                self.logger.warn(
+                    f"Tools {sorted(tools_not_found)} not found in group {group}"
+                )
+
+            # Remove a custom (non-default) tool group directory if there are
+            # no tools registered anymore under this group
+            if group != "default" and not any(self.tool_group_dir.iterdir()):
+                try:
+                    shutil.rmtree(self.tool_group_dir)
+                except OSError:
+                    self.logger.error(
+                        "Failed to remove group directory %s", self.tool_group_dir
+                    )
+                    errors = 1
+
+        return errors
+
+    def _clear_remotes(self, group) -> Union[int, list]:
+        errors = 0
+        tools_not_found = []
+        if self.context.remote:
             remotes = self.context.remote.split(",")
-        except Exception:
+        else:
             # We were not given any remotes on the command line, build the list
             # from the tools group directory.
             remotes = self.remote(self.tool_group_dir)
-            if not remotes:
-                # FIXME:  This is a weird case -- the group directory exists,
-                #  but contains no remotes. This is the result of clearing the
-                #  tools from a group which has already been cleared, because
-                #  the first clear does not remove the group directory.  This
-                #  behavior might be desirable for the `default` directory, but
-                #  it seems wrong for custom groups.
-                self.logger.error(f'No such group "{self.context.group}".')
-                return 1
-
-        tool_not_found = True  # Not found, yet
+            if (
+                not remotes
+                and group != "default"
+                and self.is_empty(self.tool_group_dir)
```

Comment:
These uses of `self.tool_group_dir` are icky.  It is set by the call to `verify_tool_group()` in the caller, and then accessed here sorta like a global variable.  It would be better if this value were passed in as a parameter.

Reply:
> The fact that we're directly referencing a property of a different class inside a method is Pythonic

A property of a different class, yes, but a parent class, and therefore (by inheritance) a property of _this_ class (instance)...so, yeah, _legal_, but yucky.  This is one aspect of Python which I don't think production code should depend on:  the ability to dynamically add attributes on the fly.  It makes the code hard to follow, and it's generally easy to avoid (simply by initializing the attribute in the class initializer, if nothing else).

But, in this case, the attribute is also updated on each iteration of the loop in the calling function, so it's really representing state which is local to the caller and not "global" to the class...so, it really should be a parameter to this function.

Reply:
@npalaska, no interest in addressing this?  (If not, I'll mark it resolved and get on with things.)

Reply:
Fixed